go-controller/pkg/ovn/controller/apbroute/external_controller_policy.go

```diff
@@ -3,377 +3,251 @@ package apbroute
-	cacheInfo, found := m.getNamespaceInfoFromCache(nsName)
-	if !found {
-		klog.V(5).Infof("Namespace %s not found in cache, creating", nsName)
-		cacheInfo = m.newNamespaceInfoInCache(nsName)
-	}
-	defer m.unlockNamespaceInfoCache(nsName)
-	return discrepancyFunc(nsName, routePolicy, cacheInfo)
-}
-
-func (m *externalPolicyManager) reconcilePolicyWithNamespacesAndPods(routePolicy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) error {
-
-	nss, err := m.listNamespacesBySelector(&routePolicy.Spec.From.NamespaceSelector)
-	if err != nil {
-		return err
-	}
-	targetNs := sets.New[string]()
-	for _, ns := range nss {
-		targetNs.Insert(ns.Name)
-	}
-
-	klog.V(4).Infof("Reconciling namespaces %+v for policy %s", strings.Join(targetNs.UnsortedList(), ", "), routePolicy.Name)
-	// reconcile Namespaces
-	currentNs, err := m.listNamespacesWithPolicy(routePolicy.Name)
-	if err != nil {
-		return err
-	}
-	klog.V(5).Infof("List of existing namespaces with policy %s: %s", routePolicy.Name, strings.Join(currentNs.UnsortedList(), ", "))
-	if len(currentNs.Difference(targetNs)) > 0 {
-		klog.V(4).Infof("Removing APB policy %s in namespaces %+v", routePolicy.Name, currentNs.Difference(targetNs))
-	}
-	// namespaces where the policy should not be applied anymore.
-	for nsName := range currentNs.Difference(targetNs) {
-		err = m.processPolicyDiscrepancyInNamespace(nsName, routePolicy, m.removePolicyFromNamespace)
-		if err != nil {
-			return err
+		// get existing policy and update routePolicySyncCache
+		existingPolicy, found := m.routePolicySyncCache.Load(policyName)
+		if !found {
+			if updatedPolicy == nil {
+				// nothing to do
+				return nil
+			}
+			existingPolicy = newRoutePolicyState()
+			m.routePolicySyncCache.Store(policyName, existingPolicy)
 		}
-	}
-	if len(targetNs.Difference(currentNs)) > 0 {
-		klog.V(4).Infof("Adding policy %s to namespaces %+v", routePolicy.Name, targetNs.Difference(currentNs))
-	}
-	// namespaces where the policy should now be applied
-	for nsName := range targetNs.Difference(currentNs) {
-		err = m.processPolicyDiscrepancyInNamespace(nsName, routePolicy, m.applyPolicyToNamespace)
-		if err != nil {
-			return err
+
+		// TODO delete debug logs
+		defer func() {
+			policy, _ := m.routePolicySyncCache.Load(policyName)
+			policyStr := ""
+			if policy != nil {
+				policyStr = policy.String()
+			}
+			klog.Infof("DEBUG: POLICY: sync policy %s, updated %+v", policyName, policyStr)
+		}()
+		klog.Infof("DEBUG: POLICY: update policy %s, existingPolicy %+v, updated %+v", policyName, existingPolicy.String(), updatedPolicy)
+
+		err = m.updateRoutePolicy(existingPolicy, updatedPolicy)
+		if updatedPolicy == nil {
+			m.routePolicySyncCache.Delete(policyName)
+			gwIPs = nil
+		} else {
+			for _, static := range updatedPolicy.staticGateways.Elems() {
+				gwIPs = gwIPs.Insert(static.Gateways.UnsortedList()...)
```

Comment:
Suggestion to change it to:
```
gwIPs = gwIPs.Union(static.Gateways)
```

Reply:
I like how this looks, but the problem is that `Union` will copy `gwIPs`, and `UnsortedList` will copy `static.Gateways`.
So I made a new function :D `insertSet` with 0 extra copies, lmk what you think about that.